definitions/npm/use-debounce_v9.x.x/flow_v0.83.x-/use-debounce_v9.x.x.js

```diff
@@ -0,0 +1,177 @@
+declare module 'use-debounce' {
+  declare type CallOptions = {|
+    /**
+     * Controls if the function should be invoked on the leading edge of the timeout.
+     */
+    leading?: boolean,
+    /**
+     * Controls if the function should be invoked on the trailing edge of the timeout.
+     */
+    trailing?: boolean,
+  |};
+
+  declare type Options = {|
+    ...CallOptions,
+    /**
+     * The maximum time the given function is allowed to be delayed before it's invoked.
+     */
+    maxWait?: number,
+  |};
+
+  declare type ControlFunctions = {|
+    /**
+     * Cancel pending function invocations
+     */
+    cancel: () => void,
+    /**
+     * Immediately invoke pending function invocations
+     */
+    flush: () => void,
+    /**
+     * Returns `true` if there are any pending function invocations
+     */
+    isPending: () => boolean,
+  |};
+
+  /**
+   * Subsequent calls to the debounced function `debounced.callback` return the result of the last func invocation.
+   * Note, that if there are no previous invocations it's mean you will get undefined. You should check it in your code properly.
+   */
+  declare type DebouncedState<R, T: (...args: any) => R> = {|
```

Comment:
Maybe use [`ReturnType`](https://flow.org/en/docs/types/utilities/#toc-return-type) if you want to stay close to the TS defs?

Reply:
Yes I had tried to follow it exactly, but when I did `T: (...args: any) => ReturnType<T>` it errors saying `T` is undefined, I suppose as a type variable `T` hasn't been initialised while computing the constraint of `T`. 

Recreated in [try flow](https://flow.org/try/#1N4Igxg9gdgZglgcxALlAIwIZoKYBsD6uEEAztvhgE6UYCe+JADpdhgCYowa5kA0I2KAFcAtiRQAXSkOz9sADwxgJ+NPTbYuQ3BMnTZA+Y2yU4IwRO4A6SFBIrGVDGM7c+IFkolXpUCWewUEAwhCQgRDH8wEH4hMnwROHlsNnw4KHwwSLAAC3wANyo4LFxscWQuHgMNZmwsiRSAWglaY1cq-hIAa2wJXNpG4Vxcdvdu3v7B0RxKUYMhKDBSqmbWwIq3eagoOrKSKgH0wtMMPznY7d2SfcoBiEZ-aG5G3Ix085AF-ZhsRoRehqUEiNMgSQHlSruBZxJrMcJwMhzAC+-EgGiCLWMAAIAIJYgC8WIAPAAVZBYgAUVmpVAQJHJp1oAEoCQA+LEAJV6QkoUBJa1JrNZFJZ+PZJIA3DEQPkTCQ4NAgvkAAxWABMAEYABxWZUgJFAA). So I just made another generic `R` which at least allows it to implicitly cast into.

Let me know if you got any other suggestions :)